### PR TITLE
vt_utils/cpu.py: Add a new utility for the CPU

### PR DESCRIPTION
**Pull request by nickzhq**
Signed-off-by: Houqi (Nick) Zuo [hzuo@redhat.com](mailto:hzuo@redhat.com)

### CODE REVIEW

**Comment by nickzhq** [comment]
@YongxueHong @luckyh @yanan-fu @zhencliu  please take a look, thanks!

File: virttest/vt_utilities/cpu.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        Update the check pattern to a certain order and format
+        """
+        pattern_list = sorted(re.split(",", flags.strip()))
+        pattern = r"(\b%s\b)" % pattern_list[0]
+        for i in pattern_list[1:]:
+            pattern += r".+(\b%s\b)" % i
+        return pattern
+
+    if ARCH in ('ppc64', 'ppc64le'):
+        return []  # remove -cpu and leave it on qemu to decide
```
Comment (YongxueHong) [inline virttest/vt_utilities/cpu.py:117]: We should update this part to adopt the current requirements of designing we wanted. This part of handling should be not decided by the virtualization testing layer.

**Comment by nickzhq** [comment]
@YongxueHong @luckyh @yanan-fu @zhencliu please take a look, thanks!

File: virttest/vt_utilities/cpu.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                cpu_support_model.append(cpu_model)
+
+    return cpu_support_model
+
+
+def get_cpu_models_on_ppc64():
+    """
+    Get cpu model from ppc64/ppc64le cpuinfo
+    """
+    return []  # if platform is the ppc64/ppc64le
```
Comment (YongxueHong) [inline virttest/vt_utilities/cpu.py:138]: Hi @maxujun 
Does the power platform support get the CPU mode on the host?
Thanks.

Comment (nickzhq) [reply virttest/vt_utilities/cpu.py:138]: > Hi @maxujun Does the power platform support get the CPU mode on the host? Thanks.

Hello @maxujun , can you give some comments?
Thanks!

Comment (maxujun) [reply virttest/vt_utilities/cpu.py:144]: Yes, support.

Comment (nickzhq) [reply virttest/vt_utilities/cpu.py:144]: hello @zhencliu @luckyh @yanan-fu ,
do you guys have any comments? Thanks!

Comment (maxujun) [reply virttest/vt_utilities/cpu.py:144]: Not support get cpu backward-compatibility mode list from host cpu info for power.only can get host self cpu model.

File: virttest/vt_utils/cpu.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    :return: cpu information
+    :rtype: dict
+    """
+    cmd = "lscpu | tee"
+    output = process.run(cmd, shell=True,
+                         ignore_status=True).stdout_text.splitlines()
+    return dict(map(lambda x: [i.strip() for i in x.split(":", 1)], output))
+
+
+def get_cpu_models():
```
[the remark this answers is not in the document]
Comment (zhencliu) [reply virttest/vt_utils/cpu.py]: It depends, we may introduce a lib function to get the cpu model name, i.e. Milan/Rome

snippet from lscpu output:
Model name:            AMD EPYC 7313 16-Core Processor

The use case here is feature owner may want to run tests on specific cpu, e.g. Milan (7313 here is Milan (codename)) 

cc @YongxueHong @nickzhq

[the remark this answers is not in the document]
Comment (nickzhq) [reply virttest/vt_utils/cpu.py]: Is it possible to have two independent functions in this module? One for `get_physical_cpu_model`, another for `get_cpu_codename`.

cc @YongxueHong @zhencliu

File: virttest/vt_utils/cpu.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+def get_cpu_model():
+    """
+    Get physical cpu model.
+
+    :return: cpu model
+    :rtype: string
+    """
+    cpu_model = process.run("lscpu|grep \"^Model name:\"")
+    cpu_model = cpu_model.split(":", 1)[-1].lstrip()
+    return cpu_model
```
Comment (YongxueHong) [inline virttest/vt_utils/cpu.py:43]:
```
>>> from virttest.vt_utils import cpu
>>> cpu.get_cpu_model()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yhong/kar/workspace/avocado-vt/virttest/vt_utils/cpu.py", line 41, in get_cpu_model
    cpu_model = process.run("lscpu|grep \"^Model name:\"")
...
FileNotFoundError: [Errno 2] No such file or directory (lscpu|grep "^Model name:"): 'lscpu|grep'
```
Looks like it does not work.

**Comment by nickzhq** [comment]
@YongxueHong @zhencliu @yafu-1  Please take a look, thanks!

**Comment by nickzhq** [comment]
@YongxueHong @zhencliu @yafu-1 Please take a look, thanks!

File: virttest/vt_utils/cpu.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    :return: cpu information
+    :rtype: dict
+    """
+    cpu_info = dict()
+    with open("/proc/cpuinfo") as fd:
+        cpu_info_lines = fd.readlines()
+    for line in cpu_info_lines:
+        if line == "\n":
+            break
```
Comment (zhencliu) [inline virttest/vt_utils/cpu.py:33]: hi @nickzhq , I am sorry that I may make a wrong comment about cpu info in the afternoon. For this function, we may return a list of dicts, in which each dict contains the information of one processor. Why? Because these information may be used for some purposes, e.g. to get the count of sockets, with shell, we can do it in this way: grep 'physical id' /proc/cpuinfo | sort | uniq | wc -l, i.e. we need all cpus' information.

or we can depend on lscpu to get the count of sockets. cc @YongxueHong

File: virttest/vt_utils/cpu.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Return information about the CPU information.
+
+    :return: cpu information
+    :rtype: dict
+    """
+    cpu_info = dict()
+    with open("/proc/cpuinfo") as fd:
+        cpu_info_lines = fd.readlines()
+    for line in cpu_info_lines:
```
Comment (zhencliu) [inline virttest/vt_utils/cpu.py:31]: for line in fd:

**Comment by nickzhq** [comment]
@YongxueHong @zhencliu @yafu-1 Please take a look, thanks!

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by nickzhq** [comment]
Hello @zhencliu @yanan-fu , do you guys have any comments? Thanks!

**Review by zhencliu — APPROVED** [review]

**Comment by nickzhq** [comment]
@luckyh @yanan-fu Please take a look, thanks!

**Comment by zhencliu** [comment]
hi @nickzhq , would you please add two functions:
get_cpu_family
get_cpu_stepping

Both are being used by sevctl command in future, cc @YongxueHong

File: virttest/vt_utils/cpu.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    Return the name of cpu family.
+
+    :return: the name of the cpu family in string
+    :rtype: string
+    """
+    cpu_family_re = "cpu family\s+:\s+(\w+)"
+    with open("/proc/cpuinfo") as fd:
+        cpu_info = fd.read()
+    cpu_family = re.findall(cpu_family_re, cpu_info)
```
Comment (zhencliu) [inline virttest/vt_utils/cpu.py:116]: we may also need to check if cpu_family is empty and return an OSError if it is, like what we did in other functions

File: virttest/vt_utils/cpu.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+
+def get_cpu_family():
+    """
+    Return the name of cpu family.
+
+    :return: the name of the cpu family in string
+    :rtype: string
+    """
+    cpu_family_re = "cpu family\s+:\s+(\w+)"
```
Comment (zhencliu) [inline virttest/vt_utils/cpu.py:113]: do you think 'c' in 'cpu family' could be a capital letter?

Please also keep an eye on stepping

File: virttest/vt_utils/cpu.py (hunk 7 above)
Comment (nickzhq) [reply virttest/vt_utils/cpu.py:116]: @zhencliu  After I made a test, the exception `OSError` should be added at here.

File: virttest/vt_utils/cpu.py (hunk 8 above)
Comment (nickzhq) [reply virttest/vt_utils/cpu.py:113]: > do you think 'c' in 'cpu family' could be a capital letter?
> 
> Please also keep an eye on stepping

After I made check, the 'C' is NOT necessary at here since the x86_64 platform only has `cpu family`.

File: virttest/vt_utils/cpu.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; specifically version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat (c) 2023
```
Comment (YongxueHong) [inline virttest/vt_utils/cpu.py:14]:
```suggestion
# Copyright: Red Hat (c) 2023, The Avocado Authors.
```

**Comment by nickzhq** [comment]
@YongxueHong @zhencliu Please take a look, thanks!

**Comment by yanan-fu** [comment]
Suggest using the MULTILINE mode for regex.

**Comment by nickzhq** [comment]
> Suggest using the MULTILINE mode for regex.

Done. Please check again, thanks! @yanan-fu

File: virttest/vt_utils/cpu.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+def get_cpu_flags():
+    """
+    Return a list of the CPU flags.
+
+    :return: cpu flags
+    :rtype: list[string]
+    :raises: An OSError will be raised if it's NOT available on the platform
+    """
+    cpu_flags_re = "flags\s+:\s+([\w\s]+)\n"
```
[the remark this answers is not in the document]
Comment (nickzhq) [reply virttest/vt_utils/cpu.py]: `vmx flags` will be added. @yanan-fu

**Comment by nickzhq** [comment]
@zhencliu @YongxueHong @yanan-fu Please take a look, thanks!

**Comment by nickzhq** [comment]
@zhencliu @YongxueHong @yanan-fu Please take a look, thanks!

File: virttest/vt_utils/cpu.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yanan-fu) [reply virttest/vt_utils/cpu.py]: I am not requesting for `vmx flags`.  I mean that the initial `get_cpu_flags` include both `flags` and `vmx flags`  when `(?m)` was not used before.

For `vmx flags`,  no usage AFAIK, and i think it is not suitable to be added in this file. @zhencliu @YongxueHong What's your opinion ?

[the remark this answers is not in the document]
Comment (zhencliu) [reply virttest/vt_utils/cpu.py]: Agreed, currently no need to add vmx flags. To get the flags of cpu, as yanan mentioned, we may need to add ^ to match the start of the string, or both flags and vms flags will be matched with re.M mode

**Comment by YongxueHong** [comment]
LGTM

**Comment by yanan-fu** [comment]
The function definition LGTM.

Just for the implementation about the `/proc/cpuinfo`,  there are several usage like:
```
results = re.findall(***)
...
results[0]
```
No union set with all the matched information,  It assume the info for different processor is same. It is ok.
Based on this,  how about return with the first matched info directly ?

**Comment by nickzhq** [comment]
> The function definition LGTM.
> 
> Just for the implementation about the `/proc/cpuinfo`, there are several usage like:
> 
> ```
> results = re.findall(***)
> ...
> results[0]
> ```
> 
> No union set with all the matched information, It assume the info for different processor is same. It is ok. Based on this, how about return with the first matched info directly ?

In my opinion, using re.findall or re.search() is okay to me. No matter the function we use, there does NOT have a significant reason to change.
@zhencliu @YongxueHong  what's your opinions?  Thanks!
cc @yanan-fu

**Comment by zhencliu** [comment]
> > The function definition LGTM.
> > Just for the implementation about the `/proc/cpuinfo`, there are several usage like:
> > ```
> > results = re.findall(***)
> > ...
> > results[0]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > No union set with all the matched information, It assume the info for different processor is same. It is ok. Based on this, how about return with the first matched info directly ?
> 
> In my opinion, using re.findall or re.search() is okay to me. No matter the function we use, there does NOT have a significant reason to change. @zhencliu @YongxueHong what's your opinions? Thanks! cc @yanan-fu

It's OK for me for the current implementation, thanks.

**Comment by YongxueHong** [comment]
> n my opinion, using re.findall or re.search() is okay to me. No matter the function we use, there does NOT have a significant reason to change.
> @zhencliu @YongxueHong what's your opinions? Thanks!
> cc @yanan-fu

Hi @nickzhq  and @yanan-fu 
Both of them are okay for me, but I prefer @yanan-fu 's suggestion since it is not ambiguous.
With the background, we know there is and only one result of the output, so we just get only one,
but if we use the `findall` to get the result, it will bring a misunderstanding that we can get multiple results here just get the first one as we expect. BTW, there is a difference of the time complexity between the `re.findall` and `re.search()` under the lots of data invoking, the `search` is faster than `findall`.
In short, it will bring a different understanding of code with different implementations sometimes. For clean code, we suggest using the appropriate code.

File: virttest/vt_utils/cpu.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    :rtype: list[string]
+    :raises: An OSError will be raised if it's NOT available on the platform
+    """
+    cpu_flags_re = "(?m)^flags\s+:\s+([\w\s]+)$"
+    with open("/proc/cpuinfo") as fd:
+        cpu_info = fd.read()
+    cpu_flag = re.search(cpu_flags_re, cpu_info)
+    if not cpu_flag:
+        raise OSError("The cpu flags were NOT found!")
+    cpu_flag = cpu_flag.groups()[0]
```
Comment (yanan-fu) [inline virttest/vt_utils/cpu.py:71]: `cpu_flags` ?

Comment (nickzhq) [reply virttest/vt_utils/cpu.py]: > `cpu_flags` ?

Done.

**Review by yanan-fu — APPROVED** [review]
Ack

**Comment by nickzhq** [comment]
Hello @YongxueHong , can you help to merge this one? thanks!

**Comment by YongxueHong** [comment]
Thanks all, let's merge this patch.

File: virttest/vt_utils/cpu.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    :rtype: string
+    :raises: An OSError will be raised if it's NOT available on the platform
+    """
+    cpu_model = process.run("lscpu").stdout_text
+    cpu_model_re = "(?m)^[mM]odel name:.*$"
+    cpu_model = re.search(cpu_model_re, cpu_model)
+    if not cpu_model:
+        raise OSError("The cpu model name was NOT found!")
+    cpu_model = cpu_model.groups()[0]
+    return cpu_model.split(":", 1)[-1].strip()
```
Comment (YongxueHong) [inline virttest/vt_utils/cpu.py:54]: Hi @nickzhq 
I hit an error when calling this function.
The following is the detail:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yhong/kar/workspace/avocado-vt/virttest/vt_utils/cpu.py", line 57, in get_cpu_model_name
    cpu_model = cpu_model.groups()[0]
IndexError: tuple index out of range
```
Could help to check it?
Thanks in advance.

Comment (nickzhq) [reply virttest/vt_utils/cpu.py:54]: > get_cpu_model_name

This bug will be fixed at patch#4030 ( https://github.com/avocado-framework/avocado-vt/pull/4030 ). Thanks!

Houqi
